File: leetcode/validParentheses.js
const map = {
    "[" : "]",
    "(" : ")",
    "{" : "}"
};

const validParentheses = str => {
    const stack = [];
    for (let i = 0; i < str.length; i++) {
        const char = str[i];
        if (map[char] !== undefined) {
            stack.push(map[char]);
        } else if (stack.pop() !== char) {
            return false;
        }
    }
    return stack.length === 0;
}

File: leetcode/validParentheses.java
public class Solution {
    public boolean isValid(String s) {
        HashMap<Character, Character> map = new HashMap<>();
        Stack<Character> stack = new Stack<>();
        char l1 = '(', r1 = ')', l2 = '[', r2 = ']', l3 = '{', r3 = '}';
        map.put(l1, r1);
        map.put(l2, r2);
        map.put(l3, r3);
        for (char c: s.toCharArray()) {
            if (map.containsKey(c)) {
                stack.push(map.get(c));
            } else if (stack.isEmpty() || stack.pop() != c) {
                return false;
            }
        }
        return stack.isEmpty();
    }
}
